test/test_utils/ec2.py

```diff
@@ -406,6 +408,150 @@ def launch_instances_with_retry(
+    return response.get("Instances")
+
+
+def launch_efa_with_reservations(
+    ec2_client, ec2_instance_type, reservations, ec2_run_instances_definition, fn_name=""
+):
+    while reservations:
+        reservation = reservations.pop(0)
+        ec2_run_instances_definition["CapacityReservationSpecification"] = {
+            "CapacityReservationTarget": {
+                "CapacityReservationId": reservation["CapacityReservationId"]
+            }
+        }
+        try:
+            instances = launch_efa(
+                ec2_client,
+                ec2_instance_type,
+                ec2_run_instances_definition,
+                reservation["AvailabilityZone"],
+            )
+            if instances:
+                LOGGER.info(
+                    f"Your EFA reservation is ready for {fn_name}, please wait to be seated. Launching..."
+                )
+                if is_mainline_context():
+                    LOGGER.info(f"Launched EFA enabled instance for {fn_name} via {reservation}")
+                return instances
+        except ClientError as e:
+            LOGGER.debug(
+                f"Failed to launch EFA instance for {fn_name} from reservation due to {e}\n"
+                "Checking additional open reservations..."
+            )
+    return []
+
+
+def validate_efa_instance_conditions(instances, minimum_number_of_instances):
+    if len(instances) == minimum_number_of_instances:
+        return True
+    if len(instances) > minimum_number_of_instances:
+        raise RuntimeError(
+            f"Launched too many instances somehow, raising and cleaning up - {instances}; min/max_allowed = {minimum_number_of_instances}"
+        )
+    return False
+
+
+class HeterogenousReservationError(Exception):
+    pass
+
+
+def launch_efa_with_heterogenous_reservations(
+    ec2_client, ec2_instance_type, ec2_run_instances_definition, fn_name=""
+):
+    minimum_number_of_instances = ec2_run_instances_definition["MinCount"]
+    # Reset max and min count to 1
+    ec2_run_instances_definition["MaxCount"] = 1
+    ec2_run_instances_definition["MinCount"] = 1
+
+    tmp_reservations = get_available_reservations(
+        ec2_client=ec2_client,
+        instance_type=ec2_run_instances_definition["InstanceType"],
+        min_availability=ec2_run_instances_definition["MinCount"],
+    )
+
+    az_counter = Counter(reservation["AvailabilityZone"] for reservation in tmp_reservations)
+    most_common_azs = [c[0] for c in az_counter.most_common()]
+
+    for most_common_az in most_common_azs:
+        reservations = [
+            reservation
+            for reservation in tmp_reservations
+            if reservation["AvailabilityZone"] == most_common_az
+        ]
+
+        available_instances = sum(
+            [reservation["AvailableInstanceCount"] for reservation in reservations]
+        )
+
+        instances = []
+        try:
+            while available_instances and len(instances) < minimum_number_of_instances:
```

Comment:
if instance launch takes about 10 min, will `available_instances` change in the meantime?

Reply:
yeah good point, will see if it is straightforward to recalculate this every time we try to launch an instance

Reply:
refreshing reservations and available instances in every az loop and every launch try